src/upload/index.ts

```diff
@@ -0,0 +1,290 @@
+import {
+  GetObjectCommand,
+  HeadObjectCommand,
+  NoSuchKey,
+  NotFound,
+  PutObjectCommand,
+  S3Client,
+} from "@aws-sdk/client-s3";
+import zlib from "zlib";
+import fs from "fs";
+
+import { LightBlockCache, lightBlockCache } from "@/cache";
+import { LightBlock } from "@/models/lightstreamer";
+import { logger } from "@/utils/logger";
+
+class UploadError extends Error {}
+
+export type BlockRange = {
+  start: number;
+  end: number;
+};
+
+export type BlockFile = {
+  blocks: string;
+  manifest: string;
+  timestamp: number;
+};
+
+export class LightBlockUpload {
+  private cache: LightBlockCache;
+  private s3Client: S3Client;
+  private chunkSizeMb: number;
+  private bucket: string;
+  private latestPath = "latest.json";
+  private blockFileName = "blocks";
+
+  constructor(cache: LightBlockCache) {
+    this.cache = cache;
+    if (!process.env["BUCKET_ENDPOINT"]) {
+      throw new UploadError("BUCKET_ENDPOINT not set");
+    }
+    if (!process.env["BUCKET_ACCESS_KEY_ID"]) {
+      throw new UploadError("BUCKET_ACCESS_KEY_ID not set");
+    }
+    if (!process.env["BUCKET_SECRET_ACCESS_KEY"]) {
+      throw new UploadError("BUCKET_SECRET_ACCESS_KEY not set");
+    }
+    if (!process.env["BUCKET_NAME"]) {
+      throw new UploadError("BUCKET_NAME not set");
+    }
+    if (!process.env["UPLOAD_CHUNK_SIZE_MB"]) {
+      throw new UploadError("UPLOAD_CHUNK_SIZE_MB not set");
+    }
+
+    this.chunkSizeMb = parseInt(
+      process.env["UPLOAD_CHUNK_SIZE_MB"] as string,
+      10,
+    );
+    this.bucket = process.env["BUCKET_NAME"];
+    this.s3Client = new S3Client({
+      region: "auto",
+      endpoint: process.env["BUCKET_ENDPOINT"],
+      credentials: {
+        accessKeyId: process.env["BUCKET_ACCESS_KEY_ID"],
+        secretAccessKey: process.env["BUCKET_SECRET_ACCESS_KEY"],
+      },
+    });
+  }
+
+  async upload(): Promise<void> {
+    const latestJson = await this.getObject(this.latestPath);
+
+    let currentUploadSize = 0;
+    if (!latestJson) {
+      console.warn("No latest json, starting upload from beginning...");
+    } else {
+      const latest: BlockFile = JSON.parse(latestJson);
```

Comment:
There is the possibility that this parsing will fail, and causing an exception, causing the `upload()` function to return prematurely without calling a new instance of itself.

I think the safest way to make sure that `upload()` runs no matter what is something like this:

```ts
async upload(): Promise<void> {
  while (true) {
    try {
      await this.uploadInner();
    } catch {
      /* ... */
    }
  }
}

private async uploadInner(): Promise<void> {
  /* ... logic ... */
}
```

or equivalently:

```ts
async upload(): Promise<void> {
  try {
    await this.uploadInner();
  } catch {
    /* ... */
  }
  return this.upload();
}

private async uploadInner(): Promise<void> {
  /* ... logic ... */
}
```

There may be other ways as well. The point is: I'd wrap _all_ the logic in a big `try`/`catch` block to ensure that *nothing* can cause the loop to stop. Especially because, if `upload()` dies, there's no way to tell from the outside (you can tell that blocks are no longer being uploaded, but it won't be immediately clear why)